### PR TITLE
adding model_builder intro notebook

### PR DESCRIPTION
**Pull request by michaelraczycki**
New notebook that shows features introduced by ModelBuilder in context of the DelayedSaturatedMMM

### CODE REVIEW

**Comment by michaelraczycki** [comment]
<p>Would you mind including the csv in the repo? I can run this code and save in csv format</p>

---

 <a href='https://app.reviewnb.com/pymc-labs/pymc-marketing/pull/324/discussion/#comment-1651189390'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

**Comment by michaelraczycki** [comment]
<p>it also looks strange for me, that's why sometimes you can see me actually reversing the autocorrect and using initialize :)</p>

---

 <a href='https://app.reviewnb.com/pymc-labs/pymc-marketing/pull/324/discussion/#comment-1651189401'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

**Comment by michaelraczycki** [comment]
<p>that's me being polish that comes out, we pack everything in the noun (like a natural metadata) so in polish we don't have such constructs like 'a' 'an' 'the' etc. So thanks for doing such a thorough check, in the future I'll try to pass my writing through some AI to catch those</p>

---

 <a href='https://app.reviewnb.com/pymc-labs/pymc-marketing/pull/324/discussion/#comment-1651189451'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

**Comment by twiecki** [comment]
<p>No, we should include it.</p>

---

 <a href='https://app.reviewnb.com/pymc-labs/pymc-marketing/pull/324/discussion/#comment-1651189390'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

**Comment by twiecki** [comment]
<p>We use American English.</p>

---

 <a href='https://app.reviewnb.com/pymc-labs/pymc-marketing/pull/324/discussion/#comment-1651189401'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

**Comment by michaelraczycki** [comment]
<p>removed :)</p>

---

 <a href='https://app.reviewnb.com/pymc-labs/pymc-marketing/pull/324/discussion/#comment-1651189375'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

**Comment by michaelraczycki** [comment]
<p>done</p>

---

 <a href='https://app.reviewnb.com/pymc-labs/pymc-marketing/pull/324/discussion/#comment-1651189425'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

**Comment by michaelraczycki** [comment]
<p>fixed</p>

---

 <a href='https://app.reviewnb.com/pymc-labs/pymc-marketing/pull/324/discussion/#comment-1651189469'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

**Comment by michaelraczycki** [comment]
<p>fixed as well</p>

---

 <a href='https://app.reviewnb.com/pymc-labs/pymc-marketing/pull/324/discussion/#comment-1651189489'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

**Comment by ricardoV94** [comment]
We have csv datasets here for the docs: https://github.com/pymc-labs/pymc-marketing/blob/main/datasets/mmm_example.csv

You can import them over the web with pandas, by just passing the `raw` link

File: docs/source/notebooks/ModelBuilder in PyMC-Marketing context.ipynb [hunk 1]
```diff
@@ -0,0 +1,1149 @@
+{
```
Comment (ricardoV94) [inline docs/source/notebooks/ModelBuilder in PyMC-Marketing context.ipynb:1]: <p>I think this intro paragraph can be made shorter and stronger. No need to be shy</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/pymc-labs/pymc-marketing/pull/324/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='1'/>

Comment (ricardoV94) [inline docs/source/notebooks/ModelBuilder in PyMC-Marketing context.ipynb:1]: <p>No need to justify why we picked the MMM as an example. Just say we pick it, but it works for all pymc MMM and CLV models.</p><p> </p><p>Also you refer to sklearn here and scikit-learn below. Should pick one.</p><p> </p><p>The <code>ModelBuilder</code> can be referenced just at the end of the article, for those interested in the internals, and should include a link to the pymc-experimental docs/ example. </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/pymc-labs/pymc-marketing/pull/324/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='3'/>

Comment (ricardoV94) [inline docs/source/notebooks/ModelBuilder in PyMC-Marketing context.ipynb:1]: <p>Now that we dropped 3.8 in PyMC, you can use dictionary update with <code>|</code>:</p><p><a href="https://docs.python.org/3/whatsnew/3.9.html#dictionary-merge-update-operators" target="_blank">https://docs.python.org/3/whatsnew/3.9.html#dictionary-merge-update-operators</a></p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/pymc-labs/pymc-marketing/pull/324/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='18'/>

Comment (ricardoV94) [inline docs/source/notebooks/ModelBuilder in PyMC-Marketing context.ipynb:1]: <p>I would show doing posterior predictive on new data after loading the model. That's a more tangible application</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/pymc-labs/pymc-marketing/pull/324/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='43'/>

Comment (carlangastr) [reply docs/source/notebooks/ModelBuilder in PyMC-Marketing context.ipynb:1]: <p>Is it possible to control the beta and alphas of beta distributions? In case the user needs to send additional information? + If so, this values should be scaled as well, right? or could be passed on real scale?</p><p> </p><p>Example:</p><pre class="ql-syntax" spellcheck="false">{'beta_channel': {'sigma': prior_sigma, 'alpha':prior_alphas, 'beta':prior_betas, 'dims': ('channel',)}}
</pre>

Comment (michaelraczycki) [reply docs/source/notebooks/ModelBuilder in PyMC-Marketing context.ipynb:1]: <p>the DelayedSaturatedMMM doesn't implement beta prior, there's alpha, which priors can be controlled by changing their values in model_config.</p>

Comment (michaelraczycki) [reply docs/source/notebooks/ModelBuilder in PyMC-Marketing context.ipynb:1]: <p>The notebook was meant to be an introduction of ModelBuilder as a module, so I'm not sure if I agree that it should be just a mention at the end (and this was supposed to be an example for it)</p>

Comment (twiecki) [reply docs/source/notebooks/ModelBuilder in PyMC-Marketing context.ipynb]: <p>Agreed.</p>

**Comment by michaelraczycki** [comment]
<p>Oh, good catch. I've intentionally left it out here but forgot to comment on that. Anything you put here, will not be persisted in the model, so It's still possible to add something to the fit kwargs, just to play around if you want to, and don't need to bother re-running it, as only the sampler_config gets saved</p>

---

 <a href='https://app.reviewnb.com/pymc-labs/pymc-marketing/pull/324/discussion/#comment-1661612589'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

**Comment by michaelraczycki** [comment]
<p>I'll check that</p><p> </p>

---

 <a href='https://app.reviewnb.com/pymc-labs/pymc-marketing/pull/324/discussion/#comment-1661612603'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

File: docs/source/notebooks/ModelBuilder in PyMC-Marketing context.ipynb (hunk 1 above)
Comment (ricardoV94) [reply docs/source/notebooks/ModelBuilder in PyMC-Marketing context.ipynb]: <p>The notebook is quite short so I think it's fine, but PyMC-Marketing users don't need to know something called <code>ModelBuilder</code> exists. It's one more piece of cognitive load that doesn't gain them anything. All they care is that PyMC-Marketing models can be configed, saved and loaded imo</p>

**Comment by ricardoV94** [comment]
Perhaps rename the jupyter notebook filename to something more succinct like `model_deployment`

File: docs/source/notebooks/ModelBuilder in PyMC-Marketing context.ipynb (hunk 1 above)
Comment (michaelraczycki) [reply docs/source/notebooks/ModelBuilder in PyMC-Marketing context.ipynb]: <p>Do we have another dataset that could be used to get the new predictions? Otherwise the only approach I see is to use the code that was used to generate this data, and change the random seed, therefore getting slightly different data, is that what you had in mind?</p><p> </p>

**Comment by michaelraczycki** [comment]
<p>done</p>

---

 <a href='https://app.reviewnb.com/pymc-labs/pymc-marketing/pull/324/discussion/#comment-1661612554'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

**Comment by michaelraczycki** [comment]
<p>done</p>

---

 <a href='https://app.reviewnb.com/pymc-labs/pymc-marketing/pull/324/discussion/#comment-1661612575'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

**Comment by michaelraczycki** [comment]
<p>netcdf is the only format that can be used by calling to_netcdf from Arviz, so I've added .nc extension to the file. I'm not sure if this actually changes anything as the file was loaded correctly before, but probably it's better to use it than not</p>

---

 <a href='https://app.reviewnb.com/pymc-labs/pymc-marketing/pull/324/discussion/#comment-1661612603'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

**Comment by michaelraczycki** [comment]
<p>removed</p>

---

 <a href='https://app.reviewnb.com/pymc-labs/pymc-marketing/pull/324/discussion/#comment-1661709138'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

**Comment by michaelraczycki** [comment]
<p>removed, my mistake overlooked it after the previous debugging</p>

---

 <a href='https://app.reviewnb.com/pymc-labs/pymc-marketing/pull/324/discussion/#comment-1661709151'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

**Comment by michaelraczycki** [comment]
<p>removed</p>

---

 <a href='https://app.reviewnb.com/pymc-labs/pymc-marketing/pull/324/discussion/#comment-1661709176'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

File: docs/source/notebooks/ModelBuilder in PyMC-Marketing context.ipynb (hunk 1 above)
Comment (michaelraczycki) [reply docs/source/notebooks/ModelBuilder in PyMC-Marketing context.ipynb]: <p>how do you like the new hopefully-not-so-shy version? :)</p>

**Comment by michaelraczycki** [comment]
<p>MMM itself just has plot_posterior_predictive, and posterior_predictive as a property. I could use the model_builder predict, but it returns only means of the posterior. But you're right, I completely forgot about the sample_posterior_predictive from the model builder itself, which would work in that case</p>

---

 <a href='https://app.reviewnb.com/pymc-labs/pymc-marketing/pull/324/discussion/#comment-1661716515'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

**Comment by michaelraczycki** [comment]
<p>done</p>

---

 <a href='https://app.reviewnb.com/pymc-labs/pymc-marketing/pull/324/discussion/#comment-1661716462'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

**Comment by michaelraczycki** [comment]
<p>we are, but it has different origins, and even though the pymc-experimental has a filter included to get rid of this message, on load it still appears (previously it would appear as well on fit, save or any function that touches idata)</p>

---

 <a href='https://app.reviewnb.com/pymc-labs/pymc-marketing/pull/324/discussion/#comment-1661716489'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

**Comment by michaelraczycki** [comment]
<p>now it uses model_builder method</p>

---

 <a href='https://app.reviewnb.com/pymc-labs/pymc-marketing/pull/324/discussion/#comment-1661716515'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

**Comment by michaelraczycki** [comment]
<p>done</p>

---

 <a href='https://app.reviewnb.com/pymc-labs/pymc-marketing/pull/324/discussion/#comment-1661716545'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

**Comment by michaelraczycki** [comment]
<p>done</p>

**Comment by ricardoV94** [comment]
@michaelraczycki I pushed some changes:
1. Remove unused imports
2. Simplify code
3. Simplify some text
4. Make text less "markety" and more "documentation"

Let me know if you have any suggestions / dislike the/some changes

**Comment by michaelraczycki** [comment]
and just as a personal opinion, the ModelBuilder probably needs more in-depth documentation as we're still struggling with few limitations. Thomas mentioned that it was supposed to be used for the blog post as well, that's why I made it more in that tone, not purely as a documentation

**Comment by ricardoV94** [comment]
> and just as a personal opinion, the ModelBuilder probably needs more in-depth documentation as we're still struggling with few limitations. Thomas mentioned that it was supposed to be used for the blog post as well, that's why I made it more in that tone, not purely as a documentation

We can keep that tone for the blogpost, but as a piece of documentation it would very quickly get old and weird imo

**Comment by ricardoV94** [comment]
> @ricardoV94 I'll leave it up to you and @twiecki if you prefer this form, it's fine for me as well. When it comes to changes I'd suggest is to not pass the actual parameters to the dummy constructor, in my opinion it complicates it and just adds more code for no good reason, as after calling the default_model_config property it's discarded anyway, that's why I included just empty strings and lists to begin with

I imagine some future models may need to adjust the priors based on the parameters, so this feels like a safer way to show users how to use it. What constitutes an "empty" model may also be hard to grasp as a user.

**Comment by michaelraczycki** [comment]
Thinking about that now, what do you think of an idea to add some static function like "show_default_model_config" to completely eliminate the necessity of instantiating the dummy model? It might not be the most elegant, but neither is creating an object just to get a feeling on how you should structure one of it's input parameters

**Comment by ricardoV94** [comment]
Yeah I also feel the current way is a bit clunky. Was there a reason it was not implemented as a static method from the beginning?

**Comment by michaelraczycki** [comment]
Honestly I thought of the default_model_config as a way of a learning support only recently, before I was more aiming for people to run the model first time on default settings, and then start to adjust them for their specific purpose

**Comment by ricardoV94** [comment]
If it doesn't break anything I would suggest making it a static method then.

**Comment by michaelraczycki** [comment]
I'll make an issue for that, it will be a looot of places where it will need to be replaced, and I'll need few hours to make sure that everything still works correctly

**Review by ricardoV94 — APPROVED** [review]
